pilot/pkg/serviceregistry/kube/controller/controller.go

```diff
@@ -503,6 +510,13 @@ func (c *Controller) onNodeEvent(obj interface{}, event model.Event) error {
 		updatedNeeded = true
 		c.Lock()
 		delete(c.nodeInfoMap, node.Name)
+		// We should remove the node for all the services that
+		// had a workload there as we don't want to route traffic.
+		// For performance reasons, we do it only for gateway services
+		// whose external traffic policy is set to Local.
+		for h := range c.nodePortGwLocalServiceToWorkloadNodesMap {
```

Comment:
Only need to do this when this node is a gateway node, right? So u can do this only when the node is in `nodeInfoMap`

Reply:
Yes. Only when the node hosts some gateway with Local policy this map is filled. Added the check.